src/trainer.py

```diff
@@ -671,8 +692,130 @@ def _aggregate_results(self, tasks, task_infos, metric_infos):
             log.info('%s, %d, %s', metric, best_epoch, all_metrics_str)
         return results
 
+    def _update_metric_history(self, epoch, all_val_metrics, metric, task_name, metric_infos, metric_decreases, should_save, new_best_macro):
+        """
+        This function updates metric history with the best validation score so far
+        Parameters
+        ---------
+        epoch: int
+        all_val_metrics: dict with current epoch's validation performance
+        metric: str, name of metric
+        task_name: str, name of task
+        metric_infos: dict storing information about the various metrics
+        metric_decreases: bool, marker to show if we should increase or
+        decrease validation metric.
+        should_save: bool, for checkpointing
+        new_best_macro: bool, indicator of whether the previous best preformance score was exceeded 
+
+        Returns
+        ________
+        metric_infos: dict storing information about the various metrics
+        this_epoch_metric: dict, metric information for this epoch, used for optimization scheduler
+        should_save: bool
+        new_best_macro: bool 
+        """
+        this_epoch_metric = all_val_metrics[metric]
+        metric_history = metric_infos[metric]['hist']
+        metric_history.append(this_epoch_metric)
+        is_best_so_far, out_of_patience = \
+            self._check_history(metric_history, this_epoch_metric, metric_decreases)
+        if is_best_so_far:
+            log.info("Best model found for %s.", task_name)
+            metric_infos[metric]['best'] = (epoch, all_val_metrics)
+            should_save = True
+            if task_name == 'macro':
+                new_best_macro = True
+        if out_of_patience:
+            metric_infos[metric]['stopped'] = True
+            log.info("Out of patience. Stopped tracking %s", task_name)
+        return metric_infos, this_epoch_metric, should_save, new_best_macro
+
+    def _get_validation_performance(self, task, task_infos, tasks, batch_size, all_val_metrics, n_examples_overall):
+        """
+        This function evaluates each task, builds validation generator, and gets the validation metrics.
```

Comment:
This doesn't quite make sense. Don't you have to build the generators in order to evaluate each task?

Reply:
Yes you do - this function does build generators and then evaluates on the task. Could you clarify what is confusing for you?

Reply:
This phrasing makes it sound like it evaluates, then builds generators, then builds metrics.